# Combining Data 1

Aggregating and summarizing data with Pandas `groupby` and `pivot_table` and merging/joining datasets using Pandas `concat` and `merge`.

## Part 1: Summarizing Data
In this part we will start by working with the `seaborn` `planets` dataset. [Seaborn](https://seaborn.pydata.org/index.html) is a library for data visualization in Python (already included in your Anaconda distribution) to which we will be returning to soon. For now, we are just using it for easy access to the `planets` dataset containing information about 1,035 extrasolar planets that have been discovered by astronomers over the last several years. As an aside, extrasolar planet discovery is an excellent example of how data science can help fuel discovery in the sciences by automating the analysis of large quantities of data, in this case from telescopes. If you're interested (not necessary to complete this assignment), you can read more at https://exoplanets.nasa.gov, from which this dataset was originally drawn.

To begin, run the following code to import the dataset into the `planets` DataFrame and preview the first five rows.

In [1]:
import seaborn as sns
import pandas as pd

planets = sns.load_dataset('planets')
planets.head()

method  number  orbital_period   mass  distance  year
0  Radial Velocity       1         269.300   7.10     77.40  2006
1  Radial Velocity       1         874.774   2.21     56.95  2008
2  Radial Velocity       1         763.000   2.60     19.84  2011
3  Radial Velocity       1         326.030  19.40    110.62  2007
4  Radial Velocity       1         516.220  10.50    119.47  2009

### Question 1
Use Pandas `groupby` operations to answer the following.

1. If you run the code `planets.groupby("number").count()` you will see different values for the different columns. Why is that?
2. What are the two `method`s that account for the most discoveries? How many discoveries were made with those `method`s?
3. In which years were more than 100 discoveries made?
4. Which `method` has found the most distant exoplanets on average (i.e., the `distance` column), and what is that average distance?
5. Which `method` has found the single most distant exoplanet in the dataset, and what `distance` is that exoplanet?

In [2]:
# 1 If you run the code planets.groupby("number").count() you will see different values for the different columns. Why is that?
print(planets.groupby("number").count())
# 2 What are the two methods that account for the most discoveries? How many discoveries were made with those methods?
print(planets.groupby("method").count())
# 3 In which years were more than 100 discoveries made?
print(planets.groupby("year").count())
# 4 Which method has found the most distant exoplanets on average (i.e., the distance column), and what is that average distance?
print(planets.groupby("method").mean())
# 5 Which method has found the single most distant exoplanet in the dataset, and what distance is that exoplanet?
print(planets.groupby("method").max())

        method  orbital_period  mass  distance  year
number                                              
1          595             555   307       513   595
2          259             257   119       158   259
3           88              87    52        66    88
4           32              32    13        20    32
5           30              30     4        20    30
6           24              24    18        24    24
7            7               7     0         7     7
                               number  orbital_period  mass  distance  year
method                                                                     
Astrometry                          2               2     0         2     2
Eclipse Timing Variations           9               9     2         4     9
Imaging                            38              12     0        32    38
Microlensing                       23               7     0        10    23
Orbital Brightness Modulation       3               3     0        

### Answer 1
1. planets.groupby("number").count() returns different values for different columns because some columns are missing data and other columns are not. Note that method and year have identical values along their respective columns as they contain complete data.  
2. The two methods of that account for the most discoveries are 1. Radial Velocity (553) and 2. Transit (397).
3. The years in which more than 100 discoveries were made are 2010, 2011, 2012, 2013.
4. Microlensing (4144).
5. Transit (8500).

### Question 2
Next we will work with the titanic dataset which contains historical information about the passengers of the cruiseship *Titanic* that sank in the North Atlantic in 1912. Import the dataset and preview the first few rows below.

In [3]:
import seaborn as sns
import pandas as pd

titanic = sns.load_dataset('titanic')
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

Use Pandas `pivot_table`s to answer the following. Note: What you choose to be the rows or columns of the `pivot_table`s is up to you as long as you are showing the correct groupings and values. We do recommend doing each part in it's own cell, for nicer table formatting.

1. Show the average fare paid by passengers grouped by each combination of `sex` and `class`.
2. Show the number of passengers grouped by each combination of `class` and `embark_town`.
3. Show the fraction of passengers who survived (i.e., `survived==1`) grouped by each combination of `sex`, `class`, and `embark_town`. For example, if there are four individuals of a given `sex`, `class`, and `embark_town`, and three of the four survived, the value for that combination would be `0.75`.
4. Show the average `age` and the total `fare` paid by passengers grouped by each combination of `class` and `sex`. For example, if there were just 2 passengers with `class==first` and `sex==male` aged `20` and `30` and having paid `fare`s of `50` and `70` each, then the average age of that combination would be `25` and the total `fare` paid would be `120`.

In [4]:
# 1 Show the average fare paid by passengers grouped by each combination of sex and class.
print(titanic.pivot_table("fare", index = "sex", columns = "class"))
# 2 Show the number of passengers grouped by each combination of class and embark_town.Show the number of passengers grouped by each combination of class and embark_town.
print(titanic.pivot_table("survived", index = "class", columns = "embark_town", aggfunc = "count"))
# 3 Show the fraction of passengers who survived (i.e., survived==1) grouped by each combination of sex, class, and embark_town.
print(titanic.pivot_table("survived", index = ["sex", "class"], columns = "embark_town", aggfunc = "mean"))
# 4 Show the average age and the total fare paid by passengers grouped by each combination of class and sex.
print(titanic.pivot_table(index = "class", columns = "sex", aggfunc = {"age":"mean", "fare":"sum"}))

class        First     Second      Third
sex                                     
female  106.125798  21.970121  16.118810
male     67.226127  19.741782  12.661633
embark_town  Cherbourg  Queenstown  Southampton
class                                          
First               85           2          127
Second              17           3          164
Third               66          72          353
embark_town    Cherbourg  Queenstown  Southampton
sex    class                                     
female First    0.976744    1.000000     0.958333
       Second   1.000000    1.000000     0.910448
       Third    0.652174    0.727273     0.375000
male   First    0.404762    0.000000     0.354430
       Second   0.200000    0.000000     0.154639
       Third    0.232558    0.076923     0.128302
              age                  fare           
sex        female       male     female       male
class                                             
First   34.611765  41.281386  9975.8250  82

## Part 2: Merging Data

We begin by studying four tips files included with this practice: `tips_Thur.csv`, `tips_Fri.csv`, `tips_Sat.csv`, and `tips_Sun.csv`. Each contains information about tips received by servers at a restaurant on the particular days of the week denoted by the file names (Thur for Thursday, Fri for Friday, Sat for Saturday, and Sun for Sunday). Below, we import and preview one of the datasets.

In [5]:
import pandas as pd
Thur = pd.read_csv("tips_Thur.csv")
Fri = pd.read_csv("tips_Fri.csv")
Sat = pd.read_csv("tips_Sat.csv")
Sun = pd.read_csv("tips_Sun.csv")
Thur.head()

total_bill   tip   sex smoker   time  size
0       27.20  4.00  Male     No  Lunch     4
1       22.76  3.00  Male     No  Lunch     2
2       17.29  2.71  Male     No  Lunch     2
3       19.44  3.00  Male    Yes  Lunch     2
4       16.66  3.40  Male     No  Lunch     2

### Question 3
Answer the following questions using the four tips datasets. You will need to combine (using Pandas `concat`) the datasets to answer some of the questions. Furthermore, some of the questions will require information about the day, which is only contained in the file names (you are welcome to add additional columns to the datasets if you wish).  

1. What is the average overall `total_bill` across all four days?
2. For each of the four days, what is the average `tip` for that day?
3. Create a pivot table that shows the average ratio of `tip` to `total_bill` (e.g., if a `tip` is `4` and the `total_bill` is `20`, then the ratio would be `0.2`) grouped by `sex` and day. 

In [6]:
# 1 What is the average overall total_bill across all four days?
concat_days = pd.concat([Thur, Fri, Sat, Sun], keys = ["Thur", "Fri", "Sat", "Sun"], names = ["day"])
print(concat_days.mean()["total_bill"])
# 2 For each of the four days, what is the average tip for that day?
print(concat_days.groupby("day").mean())
# 3 Create a pivot table that shows the average ratio of tip to total_bill
concat_days["ratio"] = concat_days["tip"] / concat_days["total_bill"]
concat_days.pivot_table("ratio", index = "sex", columns = "day", aggfunc = "mean")

19.785942622950827
      total_bill       tip      size
day                                 
Thur   17.682742  2.771452  2.451613
Fri    17.151579  2.734737  2.105263
Sat    20.441379  2.993103  2.517241
Sun    21.410000  3.255132  2.842105


day          Fri       Sat       Sun      Thur
sex                                           
Female  0.199388  0.156470  0.181569  0.157525
Male    0.143385  0.151577  0.162344  0.165276

### Answer 3
1. The average overall total_bill across all four days is 19.785942622950827.
2. Thur: 2.771452, Fri: 2.734737, Sat: 2.993103, Sun: 3.255132. 

### Question 4
In this question we will work with movie rating data in three different tables/dataframes. You will need to `merge` information from the different tables to answer the questions below. First we import and preview the tables.

In [7]:
import pandas as pd
users = pd.read_csv("users.csv")
users.head()

user_id  age sex  occupation
0        1   24   M  technician
1        2   53   F       other
2        3   23   M      writer
3        4   24   M  technician
4        5   33   F       other

In [8]:
ratings = pd.read_csv("ratings.csv")
ratings.head()

user_id  movie_id  rating
0      196       242       3
1      186       302       3
2       22       377       1
3      244        51       2
4      166       346       1

In [9]:
movies = pd.read_csv("movies.csv")
movies.head()

movie_id        movie_title
0         1   Toy Story (1995)
1         2   GoldenEye (1995)
2         3  Four Rooms (1995)
3         4  Get Shorty (1995)
4         5     Copycat (1995)

Answer the following. 

1. How many movies have been rated at least 100 times?
2. Which five users have given the highest average ratings? List their `user_id`s and their average `rating`s.
3. Create a pivot table that displays average `rating`s grouped by `sex` and `occupation`.

In [10]:
# 1 How many movies have been rated at least 100 times?
num_of_ratings = ratings["movie_id"].value_counts() 
print(len(num_of_ratings[num_of_ratings >= 100]))
# 2 Which five users have given the highest average ratings? List their user_ids and their average ratings.
print(ratings.groupby("user_id").mean().sort_values("rating", ascending = False))
# 3 Create a pivot table that displays average ratings grouped by sex and occupation.
user_ratings = pd.merge(users, ratings, on = "user_id")
print(user_ratings.pivot_table("rating", index = "sex", columns = "occupation", aggfunc = "mean"))

338
           movie_id    rating
user_id                      
849      323.347826  4.869565
688      529.000000  4.833333
507      490.051724  4.724138
628      452.333333  4.703704
928      341.625000  4.687500
...             ...       ...
774      417.808036  2.058036
685      487.550000  2.050000
445      554.844444  1.985185
405      732.933514  1.834464
181      786.220690  1.491954

[943 rows x 2 columns]
occupation  administrator    artist    doctor  educator  engineer  \
sex                                                                 
F                3.781839  3.347065       NaN  3.698857  3.751724   
M                3.555233  3.875841  3.688889  3.660246  3.537609   

occupation  entertainment  executive  healthcare  homemaker    lawyer  ...  \
sex                                                                    ...   
F                3.448889   3.773756    2.736021    3.27881  3.623188  ...   
M                3.440107   3.319610    3.639839    3.50000  3.741379  

### Answer 4
1. 338
2. The five users that have given the 5 highest ratings are 1. 849: 4.869565 (user_id: rating)2. 688: 4.833333, 3. 507: 4.724138, 4. 628: 4.703704, 5. 928: 4.687500. 

### Question 5
We will work with the `restaurants_a.csv` and `restaurants_b.csv` datasets for this question. Each contain five columns: `id` (a numeric index serving as a unique id, not correlated across the datasets), `name` (of the restaurant), `address` (the street address), `city`, and `type` (the type of restaurant). First we import and preview the data.

In [11]:
df_a = pd.read_csv("restaurants_A.csv")
df_a.head()

id                       name                         address  \
0   0             belvedere  the  9882 little santa monica blvd.   
1   1                  triangolo                 345 e. 83rd st.   
2   2              broadway deli               3rd st. promenade   
3   3  lettuce souprise you (at)                 3525 mall blvd.   
4   4                      otabe                  68 e. 56th st.   

            city              type  
0  beverly hills  pacific new wave  
1       new york           italian  
2   santa monica          american  
3         duluth        cafeterias  
4       new york             asian

In [12]:
df_b = pd.read_csv("restaurants_B.csv")
df_b.head()

id                  name                                address  \
0   22  indigo coastal grill                  1397 n. highland ave.   
1   54                  aqua                     252 california st.   
2   89             boulevard                          1 mission st.   
3  150  khan toke thai house                       5937 geary blvd.   
4  151           bacchanalia  3125 piedmont rd.  near peachtree rd.   

            city            type  
0        atlanta        eclectic  
1  san francisco  american (new)  
2  san francisco  american (new)  
3  san francisco            thai  
4        atlanta   international

Some, but not all, of the restaurants in the two datasets are actually the same. In this question, we would like to consider the problem of merging the datasets. Unfortunately, the `id`s do not correspond between the datasets, so there is no obvious primary key to merge on. In this question, you will explore a fuzzy matching to link the records between the two datasets. You will be asked to use the `edit_dist` function, but you do not need to implement it. An implementation is provided for you in `edit_distance.py`, and you can simply import the function below. It takes two strings as input and returns the edit distance between them.

In [13]:
from edit_distance import edit_dist

# Example of using the edit_dist function
print(edit_dist("hello", "hallo!"))

2


Answer the following.

1. First, try to perform an inner merge (the default for Pandas `merge`) on the two datasets on the `name` column. How many rows are in the resulting merged dataset? Why is this value much smaller than the sizes of `df_a` and `df_b`?
2. Next, try to perform an inner merge on the two datasets on the `city` column. How many rows are in the resulting merged dataset? Why is this value much larger than the sizes of `df_a` and `df_b`?
3. Print the names of all pairs of records (one from `df_a` and the other from `df_b`) such that the two names have edit distance of 1 or 2 (note that if two strings have edit distance 0, they are exactly the same; you do not need to print these). It is fine to use `for` loops to solve this and your code may take a second or two to run.
4. Among the names you identified in step 3, which pairs do you think are actually mispellings, and which do you think might actually be different restaurants? Explain your answer using information from other columns beside `name`.

In [14]:
# 1
print(pd.merge(df_a, df_b, how = "inner", on = "name").shape)
# 2 
print(pd.merge(df_a, df_b, how = "inner", on = "city").shape)
# 3 
names_a = df_a['name']
names_b = df_b['name']
distList = []
for a in names_a:
    for b in names_b:
        if edit_dist(a, b) == 1 or edit_dist(a, b) == 2:
            distList.append([a,b])
print(distList)

(43, 9)
(4887, 9)
[["l'orangerie", 'l orangerie'], ['indigo coast grill', 'indigo coastal grill'], ['boulavard', 'boulevard'], ['drago', 'spago'], ['felidia', 'filidia'], ['march', 'marichu'], ['mesa grill', 'sea grill'], ['uncle nicks', "uncle nick's"]]


### Answer 5
1. 43 rows, because many of the entries in the name columns of the two datasets do not match and the merge function requires exact matches. 
2. 4887 rows, because one city can appear multiple times on both columns of the two datasets, resulting in numerous combinations of matches. 
3. [["l'orangerie", 'l orangerie'], ['indigo coast grill', 'indigo coastal grill'], ['boulavard', 'boulevard'], ['drago', 'spago'], ['felidia', 'filidia'], ['march', 'marichu'], ['mesa grill', 'sea grill'], ['uncle nicks', "uncle nick's"]]
4. Mispellings: 3. [["l'orangerie", 'l orangerie'], ['indigo coast grill', 'indigo coastal grill'], ['boulavard', 'boulevard'],['felidia', 'filidia'], ['uncle nicks', "uncle nick's"]]
I believe these are mispellings because their addresses are the same. 
Different Restaurants: 3. [['drago', 'spago'], ['march', 'marichu'], ['mesa grill', 'sea grill']]
I believe thsse are different restaurants because they have different addresses.
